### PR TITLE
API version to support Metric Dataplane Subscriptions in Azure Monitor Private Link Scopes

### PR DESCRIPTION
**Pull request by qinjoshua**
# ARM (Control Plane) API Specification Update Pull Request 

## Purpose of this PR

What's the purpose of this PR? Check all that apply. This is **mandatory**!

  - [x] New API version. (Such PR should have been generated with [OpenAPI Hub](https://aka.ms/openapihub), per [this wiki doc](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).)
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix swagger quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I have reviewed the general guidance on the spec PR review process: https://aka.ms/specprreview.
- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data-plane related specifications.
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy).
- [X] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can request review from an ARM API Review board.

### ARM API changes review

- If you want for the ARM team to review this PR, you must add the `ARMReview` label. 
- The automation may automatically add the `ARMReview` label, if appropriate.  
  If this happens, proceed according to guidance given in GitHub comments also added by the automation.

### Breaking change review

If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), 
follow the process outlined in the [High-level Breaking Change Process doc](https://aka.ms/breakingchangesprocess#high-level-breaking-change-process).
      
## Getting help

- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For additional help, see https://aka.ms/azsdk/support/spectools.

### CODE REVIEW

**Comment by raosuhas** [comment]
> Hi @qinjoshua! For review efficiency consideration, when creating a new API version, it is required to place API specs of the base version in the first commit, and push new version updates into successive commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

Hey @qunjoshua.. can you please add the version of the file that contains the previous version of the privatelinkScopes_API file as a base commit as is called out here before start the ARM review ?

**Comment by raosuhas** [comment]
I see that your team does not renew all files for a new api version. This is not recommended since this will lead to SDKs only having some apis in the release. Is there a reason why you do not renew all the files ?

**Comment by qinjoshua** [comment]
Hi @raosuhas , thank you for taking a look at my PR.

I've added a base commit that just replicates the previous version of the API that I'm updating. Let me know if that looks good

I'm new to the team and this is my first Swagger docs update, but my understanding is that my Team only renews files for the APIs that we're updating in the newest version because those are the only files being change. In this case, I only plan to modify the private links API. I am not sure why this is not recommended, as other APIs will continue to work using whatever their latest version is.

Let me know if there's anything else I need to do.

**Comment by TimLovellSmith** [comment]
This one looks like you're declaring the Location header incorrectly per openapi

>❌ unacceptable kind of an object to dump [object Error] | "role":"Model Validation","url":"https://github.com/Azure/azure-rest-api-specs/blob/e7f921b92fbc7c9aff11ee68d60a25f21809809c/specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json"

**Comment by qinjoshua** [comment]
@rkmanda All required validations should be successfully completed, as previously discussed in the email thread. Please take a look at the changes and approve as soon as you can.

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (TimLovellSmith) [inline specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:40]: >Microsoft.Insights/ [](http://example.com/codeflow?start=47&length=19)

While I think this is added to fix a linter error...

I seem to remember this (Change casing of RP namespace in swagger) can also detect as a breaking change, since URIs are case sensitive... even though provider names are mainly supposed to be processed insensitively. @JeffreyRichter any real known problems with changing this?

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (TimLovellSmith) [inline specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:306]: >PrivateLinkScopeOperationStatus_Get [](http://example.com/codeflow?start=24&length=35)

Is this intended as an api consistency fix? 

I'm not sure whether we consider fixing this best practice or not.

I think you might get breaking change errors from having *reordered* the parameters here.

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (TimLovellSmith) [inline specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:895]: >"type": "object" [](http://example.com/codeflow?start=10&length=16)

Breaking change and SDK experts. Is this best practice? /cc @rkmanda 

I know while we get messages about not declaring it as 'object', we could also get some warnings or messages about changing it.

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (TimLovellSmith) [inline specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:970]: >  "x-ms-identifiers": [] [](http://example.com/codeflow?start=8&length=24)

you're overriding the old default behavior, hopefully its intentional

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (TimLovellSmith) [inline specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:1068]: >ScopedResourceProperties [](http://example.com/codeflow?start=5&length=24)

I notice that HybridCompute seems to implement a similar payload, but without the subscriptionLocation, or kind.

What 'kind' values are supposed to be allowed?

What kind of a subscription are we talking about here? An Azure Subscription?

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (TimLovellSmith) [inline specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:1086]: >type [](http://example.com/codeflow?start=11&length=4)

Are there a set of specificied enum values allowed for this? We normally have provisioningStates defined in swaggers as enums.

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (TimLovellSmith) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:1068]: Answer as I understand it from chat: Its a common contract _by convention_. I.e. we'd like it to be consistent for the sake of API consistency. /cc @rkmanda

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (TimLovellSmith) [inline specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:1073]: >description [](http://example.com/codeflow?start=11&length=11)

What are the allowed or anticipated values of 'kind'?
Is it readonly, or readwritable?

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (TimLovellSmith) [inline specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:1078]: >y [](http://example.com/codeflow?start=12&length=1)

Add x-ms-mutability if appropriate.

**Comment by TimLovellSmith** [comment]
[Please answer questions and remove ARMChangesRequested tag]

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (TimLovellSmith) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:1078]: Any reasn not to change this to an 'enum' also with two values, 'resource' and 'metrics', and document in the description whether other values will be added in the future?

**Comment by TimLovellSmith** [comment]
There seem to be some new linter errors show up now (either because of changes or because some linter issues got fixed):

❌ XmsEnumValidation | The enum types should have x-ms-enum type extension set with appropriate options.Location: Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json#L929 |  
❌ XmsEnumValidation | The enum types should have x-ms-enum type extension set with appropriate options.Location: Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json#L1089

**Comment by TimLovellSmith** [comment]
This avocado warning sounds like its caused by an accidental oversight?

>The example JSON file is not referenced from the swagger file.
readme: [specification/monitor/resource-manager/readme.md](https://github.com/Azure/azure-rest-api-specs/blob/f8a06f888090de1374326ddc2925aad65cdaed3a/specification/monitor/resource-manager/readme.md)
json: [preview/2023-06-01-preview/examples/PrivateLinkScopedResourceUpdatePlatformMetrics.json](https://github.com/Azure/azure-rest-api-specs/blob/f8a06f888090de1374326ddc2925aad65cdaed3a/specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/examples/PrivateLinkScopedResourceUpdatePlatformMetrics.json)

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (qinjoshua) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:308]: I'm not entirely sure what the operation_Id is intended for, but I will mention that this was part of the existing base commit - seems to be a common pattern across the file.

Comment (TimLovellSmith) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:308]: Please temporarily ignore this feedback from me as it can be covered by SDK breaking changes review process, calling this resolved for now. @qinjoshua

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (JeffreyRichter) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:40]: For ARM/RPs, changing URL casing is not considered a breaking change so it is fine here.
(For data plane, it is breaking.)

Comment (TimLovellSmith) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:40]: Great, thanks for clearing that up!

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (ms-zhenhua) [inline specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:1117]: >ScopedResourceProvisioningState [](http://example.com/codeflow?start=21&length=31)

Must not have duplicate name of x-ms-enum extension , make sure every x-ms-enum name unique.

[https://github.com/Azure/azure-openapi-validator/blob/main/docs/unique-xms-enum-name.md](https://github.com/Azure/azure-openapi-validator/blob/main/docs/unique-xms-enum-name.md)

Comment (ms-zhenhua) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:1117]: please resolve this comment and follow [Swagger-Suppression-Process](https://aka.ms/azsdk/pr-suppressions) to suppress [ResourceNameRestriction](https://github.com/Azure/azure-openapi-validator/blob/main/docs/resource-name-restriction.md) in https://github.com/Azure/azure-rest-api-specs/pull/24590/checks?check_run_id=17587599023 with breaking change reason. After that, please remove **ARMChangesRequested** label for another review.

**Comment by qinjoshua** [comment]
I see there are still ModelValidation issues around the casing for the enums. Please advise on how to make my enums case-insensitive - otherwise, validation issues should be resolved

**Comment by TimLovellSmith** [comment]
> I see there are still ModelValidation issues around the casing for the enums. Please advise on how to make my enums case-insensitive - otherwise, validation issues should be resolved

Doesn't it just want you to case it as 'Resource' instead of 'resource' in the examples? Would there be a problem with doing that?

**Comment by TimLovellSmith** [comment]
For Semantic I think there are also maybe one or two places where you need to fix

```
          "readonly": "true",
```

to

```
          "readonly": true,
```
?

**Comment by qinjoshua** [comment]
@TimLovellSmith since the API accepts case-insensitive kind, I don't want it to be labeled as case-sensitive in the swagger docs. Also not certain this is the case but it looks like all the other enum examples I've looked at uses uppercase definitions.

**Comment by qinjoshua** [comment]
This looks a little strange: https://github.com/Azure/azure-rest-api-specs/pull/24590/checks?check_run_id=17758414025

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (rkmanda) [inline specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:841]: >kind [](http://example.com/codeflow?start=21&length=4)

use odata filtering instead

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (rkmanda) [inline specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:1107]: >kind [](http://example.com/codeflow?start=9&length=4)

[ARMBlockingComment] Why not make use of the top level property in the envelope reserved for the same purpose?

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (qinjoshua) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:841]: I'm sorry, I'm not sure what you're talking about. What is odata filtering, and how does that differ from the current implementation?

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (qinjoshua) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:1107]: The kind property in Scoped Resources were introduced for this PR specifically for AMPLS, are different from the top-level property in the envelope, and serves a different purpose.

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (ms-henglu) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:841]: Please ref this doc: https://www.odata.org/getting-started/basic-tutorial/#queryData

Comment (qinjoshua) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:841]: Based on the doc, I assume that the comment was meant to encourage the use of odata $filter rather than replacing $filter. At the same time, I believe that this kind of system is outside the scope of this PR. We are selecting between two different options, not filtering based on a detailed query with multiple different parameters. The goal of this new endpoint is to highlight the differences between the new and the old kind, and adding complex query syntax would be less intuitive to customers.

Comment (rkmanda) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:841]: Not using odata filtering to start with may result in making the query parameters complex as more requirements get added in future. Your requirements currently may be simple right now, but over time these may change.

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (rkmanda) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:1107]: What is the different purpose and why can that purpose not be served by the top level kind property in the envelope?
When providing the answers please provide as much context as possible to avoid back and forths.

File: specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json
Comment (rkmanda) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:841]: Since you are still in preview its the perfect time for you to switch to the odata standard to set your APIs for success for the future

Comment (qinjoshua) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:841]: @rkmanda I don't believe this change is appropriate. Introducing OData filtering in the API will change the design and implementation of our service, which is not what we intended. We do not foresee this growing more complex, and we don't foresee the need for more flexible filtering options. This is just a toggle that was added to make clear to the end user that a new kind of scoped resource is introduced, and it's designed to be simple.

If you would like to discuss this further over call, I'll be happy to schedule some time to talk about it.

Comment (rkmanda) [reply specification/monitor/resource-manager/Microsoft.Insights/preview/2023-06-01-preview/privateLinkScopes_API.json:841]: The feedback we are giving is based on observation of many such patterns we have encountered in the past where a simple design grows into more complex ones and soon leads to bad customer experience. I dont think there is anything more to discuss. This is a strong recommendation but not a blocker so you may ship this

**Review by rkmanda — APPROVED** [review]
:shipit:

**Comment by kazrael2119** [comment]
please update [here](https://github.com/Azure/azure-rest-api-specs/pull/24590/files#diff-264accec1b0fe201eedacac26baaf709a64bed776778f36992c08265ac4b6165L76) to your new apiversion

**Comment by qinjoshua** [comment]
> please update [here](https://github.com/Azure/azure-rest-api-specs/pull/24590/files#diff-264accec1b0fe201eedacac26baaf709a64bed776778f36992c08265ac4b6165L76) to your new apiversion

I believe that is addressed here: https://github.com/Azure/azure-rest-api-specs/pull/24590/files#:~:text=%2D%20Microsoft.Insights/preview/2023%2D06%2D01%2Dpreview/privateLinkScopes_API.json

I don't intend to change the settings for older tags, let me know if there is something you'd like me to change

**Comment by kazrael2119** [comment]
> > please update [here](https://github.com/Azure/azure-rest-api-specs/pull/24590/files#diff-264accec1b0fe201eedacac26baaf709a64bed776778f36992c08265ac4b6165L76) to your new apiversion
> 
> @kazrael2119 I believe that is addressed here: https://github.com/Azure/azure-rest-api-specs/pull/24590/files#:~:text=%2D%20Microsoft.Insights/preview/2023%2D06%2D01%2Dpreview/privateLinkScopes_API.json
> 
> I don't intend to change the settings for older tags, let me know if there is something you'd like me to change

Do you mean in the future, there will be a composite apiversion like 2023-04-01 which contains all services with their new apiversion? So you don't want to change the setting for older version?

For JS, we have so many duplicate schemas in monitor. like these: https://github.com/azure-sdk/azure-sdk-for-js/pull/5546/files#diff-b4f88d7279e423e43b8b4d038d99dfdd2290d798e0e96815fd72ce98cf331da2R147-R152
And some of them may be related with your service and other service. Do you think it's good to fix these duplicate schemas in the next composite apiverison?
If yes, I think we can approve js breaking in this pr

**Comment by qinjoshua** [comment]
Hi @rkmanda 
This PR was not merged earlier due to a delay in a partner service implementing their changes. At this point, however, we are getting close to GA and we need to merge this PR from before. Can you reopen the pull request?

**Comment by TimLovellSmith** [comment]
@JeffreyRichter @rkmanda I think perceived version number age is probably starting to be a concern if this is for a new API - is this preview version number okay to carry on with?

/cc @qinjoshua

**Comment by JeffreyRichter** [comment]
2023-06-01-preview? Azure doesn't have any policies about the api-version date being related to an actual date. Just that newer versions always have a later date and that preview and stable versions not share the same date. So I'm OK with 2023-06-01-preview coming out now. @rkmanda , any concerns from your end?

**Comment by qinjoshua** [comment]
Hey bot, don't close the PR please - we're waiting on corresponding changes from a partner team before the rest api can be updated

**Comment by qinjoshua** [comment]
We are still waiting for response from partner team before this can be merged

**Comment by qinjoshua** [comment]
We are ready to merge

**Comment by TimLovellSmith** [comment]
> 2023-06-01-preview? Azure doesn't have any policies about the api-version date being related to an actual date. Just that newer versions always have a later date and that preview and stable versions not share the same date. So I'm OK with 2023-06-01-preview coming out now. @rkmanda , any concerns from your end?

@jeffrichter Newer how, or on what date, exactly? The reason I'm asking is that it seems like there are some newer versions than this *now*, since the PRs are not getting merged in version order. @qinjoshua Do the newer versions include these changes also?

**Comment by TimLovellSmith** [comment]
Got
>This is the latest API version for microsoft.insights/privatelinkscopes, although it might not be the latest api version in microsoft.insights overall

**Comment by JeffreyRichter** [comment]
The repo tooling is setup to work best when new api-versions build on top of the previous api-version. Our breaking change tooling leverages this, for example.
Teams should not be inserting api-versions in between other versions. I don't know how well things will work if this is attempted. At a minimum, people like you & me looking at the folder will be confused as to what is going on. 

Maybe I need more summarized context as to what is being attempted with this PR to weigh in properly?

**Comment by qinjoshua** [comment]
Hi @JeffreyRichter ,
What happened here is that this feature was finished about a year ago and was waiting for changes to be implemented and rolled out in a partner team before it could be released publicly. That turned out to be about a year-long wait; during that time, the new API version was already released, but the documentation was not. We are finally ready to roll out the REST API documentation for this new feature, which is why the new API version we're releasing is chronologically from last year.

Separately from that, the `Microsoft.Insights` provider is unique in that it is shared by multiple teams, and as a result has inconsistent version numbers across different resources. For instance, the latest version of `Microsoft.Insights/dataCollectionEndpoints` is not the latest version of `Microsoft.Insights/privateLinkScopes`. We do have an exemption for this inconsistency, since it is impossible for us to coordinate version numbers across every team that uses our namespace. As a result, even though 2023-06-01-preview is unfortunately not the latest version in `Microsoft.Insights`, it is the latest version of `Microsoft.Insights/privateLinkScopes`.

**Pull request closed without merging**
